src/lib/crypto/kimchi_backend/pasta/pallas_based_plonk.ml

```diff
@@ -66,83 +70,87 @@ module Rounds_vector = Rounds.Wrap_vector
 module Rounds = Rounds.Wrap
 
 module Keypair = Dlog_plonk_based_keypair.Make (struct
-  open Marlin_plonk_bindings
-
   let name = "pallas"
 
   module Rounds = Rounds
-  module Urs = Pasta_fq_urs
-  module Index = Pasta_fq_index
+  module Urs = Kimchi.Protocol.SRS.Fq
+  module Index = Kimchi.Protocol.Index.Fq
   module Curve = Curve
   module Poly_comm = Fq_poly_comm
   module Scalar_field = Field
-  module Verifier_index = Pasta_fq_verifier_index
-  module Gate_vector = Pasta_fq_index.Gate_vector
+  module Verifier_index = Kimchi.Protocol.VerifierIndex.Fq
+  module Gate_vector = Kimchi.Protocol.Gates.Vector.Fq
   module Constraint_system = R1CS_constraint_system
 end)
 
 module Proof = Plonk_dlog_proof.Make (struct
-  open Marlin_plonk_bindings
-
   let id = "pasta_pallas"
 
   module Scalar_field = Field
   module Base_field = Fp
 
   module Backend = struct
-    include Pasta_fq_proof
+    type t =
+      ( Kimchi.Foundations.Fp.t Kimchi.Foundations.or_infinity
+      , Kimchi.Foundations.Fq.t )
+      Kimchi.Protocol.prover_proof
+
+    include Kimchi.Protocol.Proof.Fq
 
     let verify = with_lagrange verify
 
     let batch_verify =
       with_lagranges (fun lgrs vks ts ->
           Async.In_thread.run (fun () -> batch_verify lgrs vks ts))
 
+    (** auxiliary is the witness, primary is not used *)
     let create_aux ~f:create (pk : Keypair.t) primary auxiliary prev_chals
         prev_comms =
+      (* external values contains [1, primary..., auxiliary ] *)
       let external_values i =
         let open Field.Vector in
         if i = 0 then Field.one
         else if i - 1 < length primary then get primary (i - 1)
         else get auxiliary (i - 1 - length primary)
       in
-      let w = R1CS_constraint_system.compute_witness pk.cs external_values in
-      let n = Pasta_fq_index.domain_d1_size pk.index in
-      let witness = Field.Vector.create () in
-      for i = 0 to Array.length w.(0) - 1 do
-        for j = 0 to n - 1 do
-          Field.Vector.emplace_back witness
-            (if j < Array.length w then w.(j).(i) else Field.zero)
-        done
-      done ;
-      create pk.index ~primary_input:(Field.Vector.create ())
-        ~auxiliary_input:witness ~prev_challenges:prev_chals
-        ~prev_sgs:prev_comms
+
+      (* compute witness *)
+      let computed_witness =
+        R1CS_constraint_system.compute_witness pk.cs external_values
+      in
+      let num_rows = Array.length computed_witness.(0) in
+
+      (* convert to Rust vector *)
+      let witness_cols =
+        Array.init Kimchi_backend_common.Constants.columns ~f:(fun col ->
+            let witness = Field.Vector.create () in
+            for row = 0 to num_rows - 1 do
+              Field.Vector.emplace_back witness computed_witness.(col).(row)
+            done ;
+            witness)
+      in
+      create pk.index witness_cols prev_chals prev_comms
```

Comment:
I think that's a good idea. Note: I removed them initially because we can't have them on the bindings themselves. In turn, we can't have them on module types that are associated with the bindings. Sometimes it's clear that a module type is associated with bindings because it is called "Bindings" something, sometimes it's not. I was really confused about that initially.